File: 임현규/programmers/Lv3/lv3_아이템_줍기_상.py
"""  
    그래프 구현 문제로 구현의 핵심은 2가지이다.
    1. 모든 좌표 2배로 늘리기
    2. 이동 경로 구현하기
    3. 외곽선을 인식해 최단경로 구하기

    이동 경로는 오로지 직사각형 둘레만을 건너야 한다. 사실 간단하다.
    직사각형 x1, y1, x2, y2의 경로는 for문을 통해 해당 경로의 2차원 배열 깂을 모두 건너기 가능한 값으로 바꿔주면 된다.

    외곽선을 인식하는 방법은 2가지가 동시에 충족해야한다.
    
    2번을 통해 구한 2차원 배열에서 둘레 + 모든 직사각형에 대해서 내부에 점이 존재하면 안됨

    이렇게 2가지가 충족하면 된다. 문제의 조건에서는 모든 직사각형 최대 갯수가 4개이므로 시간복잡도에 영향을 주지 않는다.
    
    모든 좌표를 2배로 늘리는 이유는 2번에서 이동 경로를 생성할 때 만약에 길이가 1이면 블로킹 할 수 없기 때문이다.
    예를 들어 (1,1) -> (2, 1) -> (2, 2) -> (1, 2) 로 가야하는 경로가 있다고 가정하자.
    만약 2번 방식으로 이동경로를 구하면 (1,1), (2,1), (2,2), (1,2) 모두 붙어 있기 때문에 우리가 원하는 ㄷ 자 방향이 아닌
    세로 방향으로 접근이 가능해진다. 이거 따로 예외처리해주는 것은 매우 귀찮고 힘든 일이므로 길이를 2배로 늘려서 계산해준다.

"""
from collections import deque
from dataclasses import dataclass


dy = [-1, 0, 1, 0]
dx = [0, -1, 0, 1]


@dataclass
class Rectangle:
    x1: int
    y1: int
    x2: int
    y2: int

    def is_include(self, x, y):
        return self.x1 < x < self.x2 and self.y1 < y < self.y2


def solution(rectangles, character_x, character_y, item_x, item_y):
    character_x *= 2
    character_y *= 2
    item_x *= 2
    item_y *= 2
    rects = []
    board = [[0] * 102 for _ in range(102)]
    create_board_rects(rectangles, rects, board)
    return bfs(rects, board, character_x, character_y, item_x, item_y)


def create_board_rects(rectangles, rects, board):
    for x1, y1, x2, y2 in rectangles:
        rects.append(Rectangle(2 * x1, 2 * y1, 2 * x2, 2 * y2))
        for x in range(x1 * 2, x2 * 2 + 1):
            for y in range(y1 * 2, y2 * 2 + 1):
                board[x][y] = 1


def bfs(rects, board, character_x, character_y, item_x, item_y):
    q = deque([(character_x, character_y, 0)])
    while q:
        x, y, cost = q.popleft()
        if x == item_x and y == item_y:
            return cost // 2
        board[x][y] = 0
        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]
            if board[nx][ny] == 1 and not is_in_rectangles(rects, nx, ny):
                q.append((nx, ny, cost + 1))
    print("목표 지점에 도달하지 못함")
    return -1

# 하나라도 rect 안에 Point가 존재하면 True
def is_in_rectangles(rects, x, y):
    for rect in rects:
        if rect.is_include(x, y):
            return True
    return False


print(solution([[1, 1, 7, 4], [3, 2, 5, 5], [4, 3, 6, 9], [2, 6, 8, 8]], 1, 3, 7, 8))

File: 임현규/programmers/Lv3/lv3_아이템줍기_상.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;


class Solution {

    private static final int LINE = 1;
    private static final int INVALID = 0;

    private final int[] dy = new int[]{0, 0, -1, 1};
    private final int[] dx = new int[]{-1, 1, 0, 0};

    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        int[][] map = new int[102][102];
        characterX *= 2;
        characterY *= 2;
        itemX *= 2;
        itemY *= 2;
        List<int[]> sizeUpRectangle = new ArrayList<>();
        drawLineAndSizeUpRectangle(sizeUpRectangle, rectangle, map);
        int answer = bfs(sizeUpRectangle, map, characterX, characterY, itemX, itemY);
        return answer / 2;
    }

    private void drawLineAndSizeUpRectangle(List<int[]> sizeUpRectangle, int[][] rectangle,
        int[][] map) {
        for (int[] rect : rectangle) {
            int minX = rect[0] * 2;
            int minY = rect[1] * 2;
            int maxX = rect[2] * 2;
            int maxY = rect[3] * 2;
            sizeUpRectangle.add(new int[]{minX, minY, maxX, maxY});
            for (int x = minX; x <= maxX; ++x) {
                for (int y = minY; y <= maxY; ++y) {
                    map[x][y] = LINE;
                }
            }
        }
    }

    private int bfs(List<int[]> sizeUpRectangle, int[][] map, int srcX, int srcY, int targetX,
        int targetY) {
        Deque<Info> q = new ArrayDeque<>();
        q.add(new Info(0, srcX, srcY));
        map[srcX][srcY] = INVALID;
        while (!q.isEmpty()) {
            Info info = q.pollFirst();
            if (info.x == targetX && info.y == targetY) {
                return info.dist;
            }
            for (int i = 0; i < 4; ++i) {
                int ny = info.y + dy[i];
                int nx = info.x + dx[i];
                if (map[nx][ny] == LINE && isOutPointFromRectangle(nx, ny, sizeUpRectangle)) {
                    map[nx][ny] = INVALID;
                    q.add(new Info(info.dist + 1, nx, ny));
                }
            }
        }
        return -1;
    }

    private boolean isOutPointFromRectangle(int nx, int ny, List<int[]> sizeUpRectangle) {
        for (int[] rect : sizeUpRectangle) {
            int minX = rect[0];
            int minY = rect[1];
            int maxX = rect[2];
            int maxY = rect[3];
            if (minX < nx && nx < maxX && minY < ny && ny < maxY) {
                return false;
            }
        }
        return true;
    }

    static class Info {

        private final int dist;
        private final int x;
        private final int y;

        public Info(int dist, int x, int y) {
            this.dist = dist;
            this.x = x;
            this.y = y;
        }
    }
}